winter/http/throttling.py

```diff
@@ -0,0 +1,101 @@
+import time
+import typing
+
+import dataclasses
+from django.core.cache import cache as default_cache
+from rest_framework.throttling import BaseThrottle
+
+from ..core import Component
+from ..core import annotate
+
+if typing.TYPE_CHECKING:  # pragma: no cover
+    from ..routing import Route
+
+
+@dataclasses.dataclass
+class ThrottlingAnnotation:
+    rate: str
+
+
+def throttling(rate: str):
+    return annotate(ThrottlingAnnotation(rate), single=True)
+
+
+class BaseRateThrottle(BaseThrottle):
+    rate_by_http_method = {}
+    cache = default_cache
+    cache_format = 'throttle_{scope}_{ident}'
+    scope = None
+
+    def allow_request(self, request, view):
+        self.rate = self._get_rate(request)
+
+        if self.rate is None:
+            return True
+
+        num_requests, duration = self.parse_rate(self.rate)
```

Comment:
Why not just parse it on constructor once?

Reply:
Done